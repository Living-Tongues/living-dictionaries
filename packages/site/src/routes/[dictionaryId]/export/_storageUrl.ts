import { firebaseConfig } from '$lib/firebaseConfig';

export function getStorageDownloadUrl(path: string) {
  const convertedPath = path.replace(/\//g, '%2F');
  return `https://firebasestorage.googleapis.com/v0/b/${firebaseConfig.storageBucket}/o/${convertedPath}?alt=media`;
}
