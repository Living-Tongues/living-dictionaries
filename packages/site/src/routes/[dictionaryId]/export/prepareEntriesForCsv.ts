import type {
  ExpandedEntry,
  IDictionary,
  IPartOfSpeech,
  ISpeaker,
} from '@living-dictionaries/types';
import { stripHTMLTags } from './stripHTMLTags';
import { friendlyName } from './friendlyName';
import {
  assign_local_orthographies_as_headers,
  assign_example_sentences_as_headers,
  assign_gloss_languages_as_headers,
  assign_semantic_domains_as_headers,
  count_maximum_semantic_domains_only_from_first_senses,
} from './assignHeadersForCsv';
import {
  find_part_of_speech_abbreviation,
  get_first_speaker_from_first_sound_file,
  display_speaker_gender,
  format_local_orthographies,
  format_semantic_domains,
  format_gloss_languages,
  format_example_sentences,
} from './assignFormattedEntryValuesForCsv';
import { decades } from '$lib/components/media/ages';

export enum StandardEntryCSVFields {
  id = 'Entry Id',
  lexeme = 'Lexeme/Word/Phrase',
  phonetic = 'Phonetic (IPA)',
  interlinearization = 'Interlinearization',
  noun_class = 'Noun class',
  morphology = 'Morphology',
  plural_form = 'Plural form',
  dialects = 'Dialects',
  notes = 'Notes',
  sources = 'Source(s)',
  parts_of_speech_abbreviation = 'Part of Speech abbreviation',
  parts_of_speech = 'Part of Speech',
  image_filename = 'Image filename',
  sound_filename = 'Audio filename',
  speaker_name = 'Speaker name',
  speaker_birthplace = 'Speaker birthplace',
  speaker_decade = 'Speaker decade',
  speaker_gender = 'Speaker gender',
}
export type EntryForCSVKeys = keyof typeof StandardEntryCSVFields;
type StandardEntryForCSV = {
  [key in EntryForCSVKeys]?: string;
};

export interface EntryForCSV extends StandardEntryForCSV {
  vernacular_example_sentence?: string;
  variant?: string; // optional for Babanki & Torwali
  sound_file_path?: string; // for downloading file, not exported in CSV
  image_file_path?: string; // for downloading file, not exported in CSV
}

const dictionaries_with_variant = ['babanki', 'torwali'];

export function prepareEntriesForCsv(
  expanded_entries: ExpandedEntry[],
  dictionary: IDictionary,
  speakers: ISpeaker[],
  global_parts_of_speech: IPartOfSpeech[]
): EntryForCSV[] {
  const max_semantic_domain_number =
    count_maximum_semantic_domains_only_from_first_senses(expanded_entries);

  const default_headers: EntryForCSV = { ...StandardEntryCSVFields };

  // Begin dynamic headers
  const local_orthographies_headers = assign_local_orthographies_as_headers(
    dictionary.alternateOrthographies
  );
  const semantic_domains_headers = assign_semantic_domains_as_headers(max_semantic_domain_number);
  const gloss_languages_headers = assign_gloss_languages_as_headers(dictionary.glossLanguages);
  const example_sentences_headers = assign_example_sentences_as_headers(
    dictionary.glossLanguages,
    dictionary.name
  );

  // Dictionary specific
  if (dictionaries_with_variant.includes(dictionary.id)) {
    default_headers['variant'] = 'Variant';
  }

  const headers = {
    ...default_headers,
    ...local_orthographies_headers,
    ...semantic_domains_headers,
    ...gloss_languages_headers,
    ...example_sentences_headers,
  };

  const formattedEntries: EntryForCSV[] = expanded_entries.map((entry) => {
    const formatted_entry = {
      id: entry.id,
      lexeme: entry.lexeme,
      phonetic: entry.phonetic,
      interlinearization: entry.interlinearization,
      noun_class: entry.senses?.[0]?.noun_class,
      morphology: entry.morphology,
      plural_form: entry.plural_form,
      dialects: entry.dialects?.[0],
      notes: stripHTMLTags(entry.notes),
      sources: entry.sources?.join(' | '),
      parts_of_speech_abbreviation: find_part_of_speech_abbreviation(
        global_parts_of_speech,
        entry.senses?.[0]?.parts_of_speech?.[0]
      ),
      parts_of_speech: entry.senses?.[0]?.parts_of_speech?.[0],
      image_filename: friendlyName(entry, entry.senses?.[0].photo_files?.[0].fb_storage_path),
      sound_filename: friendlyName(entry, entry.sound_files?.[0].fb_storage_path),
    } as EntryForCSV;

    //Begin dynamic values
    formatted_entry.image_file_path = entry.senses?.[0].photo_files?.[0].fb_storage_path;
    const speaker = get_first_speaker_from_first_sound_file(entry, speakers);
    formatted_entry.sound_file_path = entry.sound_files?.[0].fb_storage_path;
    formatted_entry.speaker_name = speaker?.displayName;
    formatted_entry.speaker_birthplace = speaker?.birthplace;
    formatted_entry.speaker_decade = decades[speaker?.decade];
    formatted_entry.speaker_gender = display_speaker_gender(speaker?.gender);

    const formatted_local_orthographies = format_local_orthographies(
      entry,
      local_orthographies_headers
    );
    const formatted_semantic_domains = format_semantic_domains(entry, max_semantic_domain_number);

    const formatted_gloss_languages = format_gloss_languages(entry, dictionary.glossLanguages);

    const formatted_example_sentences = format_example_sentences(entry, dictionary.glossLanguages);

    // Dictionary specific
    if (dictionaries_with_variant.includes(dictionary.id)) {
      formatted_entry.variant = entry.variant;
    }

    return {
      ...formatted_entry,
      ...formatted_local_orthographies,
      ...formatted_semantic_domains,
      ...formatted_gloss_languages,
      ...formatted_example_sentences,
    };
  });
  return [headers, ...formattedEntries];
}
