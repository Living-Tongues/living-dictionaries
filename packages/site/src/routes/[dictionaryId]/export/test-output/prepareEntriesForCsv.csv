Entry Id,Lexeme/Word/Phrase,Phonetic (IPA),Interlinearization,Noun class,Morphology,Plural form,Dialects,Notes,Source(s),Part of Speech abbreviation,Part of Speech,Image filename,Audio filename,Speaker name,Speaker birthplace,Speaker decade,Speaker gender,native_script_1,native_script_2,Semantic domain 1,Semantic domain 2,العَرَبِيَّة‎ Gloss,English Gloss,español Gloss,Example sentence in TestLang,Example sentence in العَرَبِيَّة‎,Example sentence in English,Example sentence in español
12345qwerty,xiangjiao,xiangjiao,,,,,dialect x,"This is an example of a note, here we can write whatever we want.",A book | www.mybook.com,n,noun,12345qwerty_foo.png,12345qwerty_foo.mp3,John Smith,Whoville,41-50,male,,کیلا,Body parts,Body functions,foo,banana,,我很喜歡吃香蕉,,This is a banana,
34qw,tree,,,,,,,,,,,,,,,,,𑃐𑃥𑃝𑃢 𑃒𑃦𑃗𑃠𑃤,চুড়া বংজি,,,,,árbol,,,,