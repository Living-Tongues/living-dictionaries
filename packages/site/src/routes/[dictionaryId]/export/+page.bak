<script lang="ts">
  import { onMount } from 'svelte'
  import { Button, ShowHide } from 'svelte-pieces'
  import type { EntryView } from '@living-dictionaries/types'
  import { getCollection } from 'sveltefirets'
  import DownloadMedia from './DownloadMedia.svelte'
  import { fetchSpeakers } from './fetchSpeakers'
  import { type EntryForCSV, formatCsvEntries, getCsvHeaders } from './prepareEntriesForCsv'
  import SeoMetaTags from '$lib/components/SeoMetaTags.svelte'
  import Progress from '$lib/export/Progress.svelte'
  import { partsOfSpeech } from '$lib/mappings/parts-of-speech'
  import { page } from '$app/stores'
  import { downloadObjectsAsCSV } from '$lib/export/csv'

  export let data
  $: ({ is_manager, dictionary, admin } = data)

  let includeImages = false
  let includeAudio = false

  const entryHeaders: EntryForCSV = {}
  const formattedEntries: EntryForCSV[] = []
  const entriesWithImages: EntryForCSV[] = []
  const entriesWithAudio: EntryForCSV[] = []

  const mounted = false

  onMount(async () => {
  // const database_entries = await getCollection<ActualDatabaseEntry>(`dictionaries/${$dictionary.id}/words`)
    // const expanded_entries = database_entries.map(entry => convert_and_expand_entry(entry, $page.data.t))
    // const speakers = await fetchSpeakers(expanded_entries)

    // entryHeaders = getCsvHeaders(expanded_entries, $dictionary)
    // formattedEntries = formatCsvEntries(expanded_entries, speakers, partsOfSpeech)
    // entriesWithImages = formattedEntries.filter(entry => entry.image_filename)
    // entriesWithAudio = formattedEntries.filter(entry => entry.sound_filename)

    // mounted = true
  })
</script>

<h3 class="text-xl font-semibold mb-4">{$page.data.t('misc.export')}</h3>
{#if $is_manager}
  <div class="mb-6">
    <div>
      <i class="far fa-check" />
      {$page.data.t('export.csv_data')})
    </div>
    <div
      class="flex items-center mt-2 {entriesWithImages.length
        ? ''
        : 'opacity-50 cursor-not-allowed'}">
      <input
        disabled={!entriesWithImages.length}
        id="images"
        type="checkbox"
        bind:checked={includeImages} />
      <label for="images" class="mx-2 block leading-5 text-gray-900">
        {$page.data.t('misc.images')} ({entriesWithImages.length})</label>
    </div>
    {#if !mounted}
      <p class="text-xs italic text-orange-400 p-2">
        {$page.data.t('export.checking_images')}
      </p>
    {:else if !entriesWithImages.length}
      <p class="text-sm text-red-700 p-3">
        {$page.data.t('export.no_images')}
      </p>
    {/if}

    <div
      class="flex items-center mt-2 {entriesWithAudio.length
        ? ''
        : 'opacity-50 cursor-not-allowed'}">
      <input id="audio" type="checkbox" bind:checked={includeAudio} />
      <label for="audio" class="mx-2 block leading-5 text-gray-900">
        {$page.data.t('entry_field.audio')} ({entriesWithAudio.length})</label>
    </div>
    {#if !mounted}
      <p class="text-xs italic text-orange-400 p-2">
        {$page.data.t('export.checking_audios')}
      </p>
    {:else if !entriesWithAudio.length}
      <p class="text-sm text-red-700 p-3">
        {$page.data.t('export.no_audios')}
      </p>
    {/if}
  </div>

  {#if includeImages || includeAudio}
    <ShowHide let:show let:toggle>
      {#if !show}
        <Button onclick={toggle} form="filled">
          {$page.data.t('export.download_csv')}
          {#if includeImages}
            + {$page.data.t('misc.images')}
          {/if}
          {#if includeAudio}
            + {$page.data.t('entry_field.audio')}
          {/if}
        </Button>
      {:else}
        <DownloadMedia
          dictionary={$dictionary}
          {entryHeaders}
          finalizedEntries={formattedEntries}
          entriesWithImages={includeImages ? entriesWithImages : []}
          entriesWithAudio={includeAudio ? entriesWithAudio : []}
          on:completed={toggle}
          let:progress>
          <Progress {progress} />
          {#if progress < 1}
            <Button onclick={toggle} color="red">{$page.data.t('misc.cancel')}</Button>
          {:else}
            <Button onclick={toggle}>{$page.data.t('misc.reset')}</Button>
          {/if}
        </DownloadMedia>
      {/if}
    </ShowHide>
  {:else}
    <Button
      loading={!formattedEntries.length}
      onclick={() => {
        downloadObjectsAsCSV(entryHeaders, formattedEntries, $dictionary.id)
      }}
      form="filled">
      {$page.data.t('export.download_csv')}
    </Button>
  {/if}
{:else}
  <p>{$page.data.t('export.availability')}</p>
{/if}

{#if $admin}
  <div class="mt-5">
    <Button form="filled" href="entries/print">{$page.data.t('export.download_pdf')}</Button>
  </div>
{/if}

<SeoMetaTags
  title={$page.data.t('misc.export')}
  dictionaryName={$dictionary.name}
  description="Dictionary managers can easily export their Living Dictionary\'s text data as a .CSV spreadsheet as well as export their images and audio files in convenient ZIP folders."
  keywords="How to print a dictionary, How to create lessons for endangered languages, Language Documentation, Language Revitalization, Build a Dictionary, Online Dictionary, Digital Dictionary, Dictionary Software, Free Software, Online Dictionary Builder, Living Dictionaries, Living Dictionary, Edit a dictionary, Search a dictionary, Browse a dictionary, Explore a Dictionary" />
