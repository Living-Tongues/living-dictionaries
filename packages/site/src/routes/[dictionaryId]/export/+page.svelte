<script lang="ts">
  import { onMount } from 'svelte';
  import { t } from 'svelte-i18n';
  import { admin, dictionary, isManager } from '$lib/stores';
  import { Button, ShowHide } from 'svelte-pieces';
  import { partsOfSpeech } from '$lib/mappings/parts-of-speech';
  import type { ActualDatabaseEntry } from '@living-dictionaries/types';
  import { getCollection } from 'sveltefirets';
  import Progress from '$lib/export/Progress.svelte';
  import SeoMetaTags from '$lib/components/SeoMetaTags.svelte';
  import { convert_entry_to_current_shape } from '$lib/transformers/convert_entry_to_current_shape';
  import { expand_entry } from '$lib/transformers/expand_entry';
  import DownloadMedia from './DownloadMedia.svelte';
  import { fetchSpeakers } from './fetchSpeakers';
  import { getCsvHeaders, formatCsvEntries, type EntryForCSV } from './prepareEntriesForCsv';
  import { downloadObjectsAsCSV } from '$lib/export/csv';

  let includeImages = false;
  let includeAudio = false;

  let entryHeaders: EntryForCSV = {}
  let formattedEntries: EntryForCSV[] = [];
  let entriesWithImages: EntryForCSV[] = [];
  let entriesWithAudio: EntryForCSV[] = [];

  let mounted = false;

  onMount(async () => {
    const database_entries = await getCollection<ActualDatabaseEntry>(`dictionaries/${$dictionary.id}/words`);
    const converted_to_current_shaped_entries = database_entries.map(convert_entry_to_current_shape);
    const expanded_entries = converted_to_current_shaped_entries.map(expand_entry);
    const speakers = await fetchSpeakers(expanded_entries);

    entryHeaders = getCsvHeaders(expanded_entries, $dictionary)
    formattedEntries = formatCsvEntries(expanded_entries, speakers, partsOfSpeech)
    entriesWithImages = formattedEntries.filter((entry) => entry.image_filename);
    entriesWithAudio = formattedEntries.filter((entry) => entry.sound_filename);

    mounted = true;
  });
</script>

<h3 class="text-xl font-semibold mb-4">{$t('misc.export', { default: 'export' })}</h3>
{#if $isManager}
  <div class="mb-6">
    <div>
      <i class="far fa-check" />
      {$t('export.csv_data', { default: 'Data as .CSV' })} ({$t('export.spreadsheet', { default: 'Spreadsheet' })})
    </div>
    <div
      class="flex items-center mt-2 {entriesWithImages.length
        ? ''
        : 'opacity-50 cursor-not-allowed'}">
      <input
        disabled={!entriesWithImages.length}
        id="images"
        type="checkbox"
        bind:checked={includeImages} />
      <label for="images" class="mx-2 block leading-5 text-gray-900">
        {$t('misc.images', { default: 'Images' })} ({entriesWithImages.length})</label>
    </div>
    {#if !mounted}
      <p class="text-xs italic text-orange-400 p-2">
        {$t('export.checking_images', { default: 'Checking if image files exist' })}
      </p>
    {:else if !entriesWithImages.length}
      <p class="text-sm text-red-700 p-3">
        {$t('export.no_images', { default: 'There are no image files' })}
      </p>
    {/if}

    <div
      class="flex items-center mt-2 {entriesWithAudio.length
        ? ''
        : 'opacity-50 cursor-not-allowed'}">
      <input id="audio" type="checkbox" bind:checked={includeAudio} />
      <label for="audio" class="mx-2 block leading-5 text-gray-900">
        {$t('entry.audio', { default: 'Audio' })} ({entriesWithAudio.length})</label>
    </div>
    {#if !mounted}
      <p class="text-xs italic text-orange-400 p-2">
        {$t('export.checking_audios', { default: 'Checking if audio files exist' })}
      </p>
    {:else if !entriesWithAudio.length}
      <p class="text-sm text-red-700 p-3">
        {$t('export.no_audios', { default: 'There are no audio files' })}
      </p>
    {/if}
  </div>

  {#if includeImages || includeAudio}
    <ShowHide let:show let:toggle>
      {#if !show}
        <Button onclick={toggle} form="filled">
          {$t('export.download_csv', { default: 'Download CSV' })}
          {#if includeImages}
            + {$t('misc.images', { default: 'Images' })}
          {/if}
          {#if includeAudio}
            + {$t('entry.audio', { default: 'Audio' })}
          {/if}
        </Button>
      {:else}
        <DownloadMedia
          dictionary={$dictionary}
          {entryHeaders}
          finalizedEntries={formattedEntries}
          entriesWithImages={includeImages ? entriesWithImages : []}
          entriesWithAudio={includeAudio ? entriesWithAudio : []}
          on:completed={toggle}
          let:progress>
          <Progress {progress} />
          {#if progress < 1}
            <Button onclick={toggle} color="red">{$t('misc.cancel', { default: 'Cancel' })}</Button>
          {:else}
            <Button onclick={toggle}>{$t('misc.reset', { default: 'Reset' })}</Button>
          {/if}
        </DownloadMedia>
      {/if}
    </ShowHide>
  {:else}
    <Button
      loading={!formattedEntries.length}
      onclick={() => {
        downloadObjectsAsCSV(entryHeaders, formattedEntries, $dictionary.id);
      }}
      form="filled">
      {$t('export.download_csv', { default: 'Download CSV' })}
    </Button>
  {/if}
{:else}
  <p>{$t('export.availability', { default: 'Export is only available to dictionary managers' })}</p>
{/if}

{#if $admin}
  <div class="mt-5">
    <Button form="filled" href="entries/print">{$t('export.download_pdf', { default: 'Download PDF' })}</Button>
  </div>
{/if}

<SeoMetaTags
  title={$t('misc.export', { default: 'export' })}
  dictionaryName={$dictionary.name}
  description={$t('', {
    default:
      'Dictionary managers can easily export their Living Dictionary\'s text data as a .CSV spreadsheet as well as export their images and audio files in convenient ZIP folders.',
  })}
  keywords="How to print a dictionary, How to create lessons for endangered languages, Language Documentation, Language Revitalization, Build a Dictionary, Online Dictionary, Digital Dictionary, Dictionary Software, Free Software, Online Dictionary Builder, Living Dictionaries, Living Dictionary, Edit a dictionary, Search a dictionary, Browse a dictionary, Explore a Dictionary" />
