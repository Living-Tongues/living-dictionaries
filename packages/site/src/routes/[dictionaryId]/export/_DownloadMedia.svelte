<script lang="ts">
  import JSZip from 'jszip';
  import { fileAsBlob } from '$lib/export/csv';
  import type { IDictionary } from '@ld/types';
  import { createEventDispatcher, onDestroy, onMount } from 'svelte';
  import { getStorageDownloadUrl } from './_storageUrl';

  export let dictionary: IDictionary;
  export let formattedEntries: any[];
  export let entriesWithImages: any[] = [];
  export let entriesWithAudio: any[] = [];

  const dispatch = createEventDispatcher<{ completed: null }>();

  let fetched = 0;
  $: progress = fetched / (entriesWithImages.length + entriesWithAudio.length);
  let errors = [];
  let destroyed = false;

  onMount(async () => {
    const zip = new JSZip();

    for (const entry of entriesWithImages) {
      if (destroyed) return;
      try {
        const response = await fetch(getStorageDownloadUrl(entry.pfpa));
        if (response.ok) {
          const blob = await response.blob();
          zip.folder(`${dictionary.id}_Images/`).file(entry.pfFriendlyName, blob, { binary: true });
        } else {
          errors = [
            ...errors,
            `Entry: ${entry.lx}, Id: ${entry.id}, File: ${entry.pfpa}, Error: ${response.statusText}`,
          ];
        }
      } catch (e) {
        errors = [...errors, `Entry: ${entry.id}, File: ${entry.pfpa}, ${e}`];
      }
      fetched++;
    }
    for (const entry of entriesWithAudio) {
      if (destroyed) return;
      try {
        const response = await fetch(getStorageDownloadUrl(entry.sfpa));
        if (response.ok) {
          const blob = await response.blob();
          zip.folder(`${dictionary.id}_Audio/`).file(entry.sfFriendlyName, blob, { binary: true });
        } else {
          errors = [
            ...errors,
            `Entry: ${entry.lx}, Id: ${entry.id}, File: ${entry.sfpa}, Error: ${response.status} ${response.statusText}`,
          ];
        }
      } catch (e) {
        errors = [...errors, `Entry: ${entry.id}, File: ${entry.sfpa}, ${e}`];
      }
      fetched++;
    }

    const finalizedEntries = formattedEntries.map((entry) => {
      const newEntry = { ...entry };
      delete newEntry.pfpa;
      delete newEntry.sfpa;
      return newEntry;
    });
    const CSVBlob = fileAsBlob(finalizedEntries);
    zip.file(`${dictionary.id}.csv`, CSVBlob);

    const { saveAs } = await import('file-saver');
    const blob = await zip.generateAsync({ type: 'blob' });
    const d = new Date();
    const date = d.getMonth() + 1 + '_' + d.getDate() + '_' + d.getFullYear();
    if (destroyed) return;
    saveAs(blob, `${dictionary.id}_${date}.zip`);
    if (!errors.length) {
      dispatch('completed');
    }
  });

  onDestroy(() => {
    destroyed = true;
  });
</script>

<div>
  <slot {progress} />
</div>

{#if errors.length}
  <div class="mt-4 text-red-600 text-sm">
    <div class="font-semibold">Errors:</div>
    {#each errors as error}
      <div class="mb-3">{error}</div>
    {/each}
  </div>
{/if}
