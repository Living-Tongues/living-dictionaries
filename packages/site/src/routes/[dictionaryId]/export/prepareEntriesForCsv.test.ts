import type {
  ExpandedEntry,
  IDictionary,
  ISpeaker,
  IPartOfSpeech,
} from '@living-dictionaries/types';
import { StandardEntryCSVFields, prepareEntriesForCsv } from './prepareEntriesForCsv';
import { objectsToCsvByHeaders } from '$lib/export/csv';

describe('prepareEntriesForCsv', () => {
  const speakers: ISpeaker[] = [
    {
      displayName: 'John Smith',
      id: '123',
      birthplace: 'Whoville',
      decade: 4,
      gender: 'm',
    },
  ];
  const partsOfSpeech: IPartOfSpeech[] = [{ enAbbrev: 'n', enName: 'noun' }]; // TODO: after updated expanded entries to include abbreviations, this will no longer be needed in prepareEntriesForCsv

  test('everything', () => {
    const dictionary: IDictionary = {
      name: 'TestLang',
      id: 'testDictionary',
      glossLanguages: ['ar', 'en', 'es'],
      alternateOrthographies: ['native_script_1', 'native_script_2'],
    };
    const entries: ExpandedEntry[] = [
      {
        id: '12345qwerty',
        lexeme: 'xiangjiao',
        local_orthography_2: 'کیلا',
        senses: [
          {
            glosses: { ar: 'foo', en: 'banana' },
            parts_of_speech: ['noun', 'adjective'],
            semantic_domains: ['Body parts', 'Body functions'],
            example_sentences: [{ en: 'This is a banana', vn: '我很喜歡吃香蕉' }],
            photo_files: [
              { fb_storage_path: 'https://database.com/image.png', uid_added_by: 'Diego' },
            ],
          },
        ],
        phonetic: 'xiangjiao',
        dialects: ['dialect x'],
        notes: 'This is an example of a note, here we can write whatever we want.',
        sources: ['A book', 'www.mybook.com'],
        sound_files: [{ fb_storage_path: 'https://database.com/sound.mp3', speaker_ids: ['123'] }],
      },
      {
        id: '34qw',
        lexeme: 'tree',
        local_orthography_1: '𑃐𑃥𑃝𑃢 𑃒𑃦𑃗𑃠𑃤',
        local_orthography_2: 'চুড়া বংজি',
        senses: [{ glosses: { es: 'árbol' } }],
      },
    ];
    const [headerRow, firstEntry, secondEntry] = prepareEntriesForCsv(
      entries,
      dictionary,
      speakers,
      partsOfSpeech
    );

    const expectedHeaders_PlusDynamic_ArEnEs_TwoLocalOrthographies = {
      ...StandardEntryCSVFields,
      local_orthography_1: 'native_script_1',
      local_orthography_2: 'native_script_2',
      ar_gloss_language: 'العَرَبِيَّة‎ Gloss',
      en_gloss_language: 'English Gloss',
      es_gloss_language: 'español Gloss',
      ar_example_sentence: 'Example sentence in العَرَبِيَّة‎',
      en_example_sentence: 'Example sentence in English',
      es_example_sentence: 'Example sentence in español',
      vernacular_example_sentence: 'Example sentence in TestLang',
      semantic_domain_1: 'Semantic domain 1',
      semantic_domain_2: 'Semantic domain 2',
    };
    expect(headerRow).toMatchObject(expectedHeaders_PlusDynamic_ArEnEs_TwoLocalOrthographies);

    expect(firstEntry).toMatchSnapshot();
    expect(secondEntry).toMatchSnapshot();

    expect(objectsToCsvByHeaders(headerRow, [firstEntry, secondEntry])).toMatchFileSnapshot('./test-output/prepareEntriesForCsv.csv');
  });

  //TODO after make everything passes, allow multiple parts of speech

  describe('variant column', () => {
    test('added to babanki', () => {
      const dictionary = { id: 'babanki', glossLanguages: [] } as IDictionary;
      const entries: ExpandedEntry[] = [
        {
          lexeme: 'foo',
          variant: 'fooey',
        },
        {
          lexeme: 'baz',
        },
      ];
      const [headerRow, firstEntry, secondEntry] = prepareEntriesForCsv(
        entries,
        dictionary,
        speakers,
        partsOfSpeech
      );

      expect(headerRow.variant).toEqual('Variant');
      expect(firstEntry.variant).toEqual('fooey');
      expect(secondEntry.variant).toEqual(undefined);
    });

    test('not added to fooDictionary', () => {
      const dictionary = { id: 'fooDictionary', glossLanguages: [] } as IDictionary;
      const entries: ExpandedEntry[] = [{ lexeme: 'foo' }];
      const [headerRow] = prepareEntriesForCsv(entries, dictionary, speakers, partsOfSpeech);
      expect(headerRow.variant).toBeFalsy();
    });
  });
});
