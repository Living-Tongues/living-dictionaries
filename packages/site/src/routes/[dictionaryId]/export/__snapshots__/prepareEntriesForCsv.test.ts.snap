// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prepareEntriesForCsv > everything 1`] = `
{
  "ar_example_sentence": undefined,
  "ar_gloss_language": "foo",
  "dialects": "dialect x",
  "en_example_sentence": "This is a banana",
  "en_gloss_language": "banana",
  "es_example_sentence": undefined,
  "es_gloss_language": undefined,
  "id": "12345qwerty",
  "image_file_path": "https://database.com/image.png",
  "image_filename": "12345qwerty_foo.png",
  "interlinearization": undefined,
  "lexeme": "xiangjiao",
  "local_orthography_1": undefined,
  "local_orthography_2": "کیلا",
  "morphology": undefined,
  "notes": "This is an example of a note, here we can write whatever we want.",
  "noun_class": undefined,
  "parts_of_speech": "noun",
  "parts_of_speech_abbreviation": "n",
  "phonetic": "xiangjiao",
  "plural_form": undefined,
  "semantic_domain_1": "Body parts",
  "semantic_domain_2": "Body functions",
  "sound_file_path": "https://database.com/sound.mp3",
  "sound_filename": "12345qwerty_foo.mp3",
  "sources": "A book | www.mybook.com",
  "speaker_birthplace": "Whoville",
  "speaker_decade": "41-50",
  "speaker_gender": "male",
  "speaker_name": "John Smith",
  "vernacular_example_sentence": "我很喜歡吃香蕉",
}
`;

exports[`prepareEntriesForCsv > everything 2`] = `
{
  "ar_example_sentence": undefined,
  "ar_gloss_language": undefined,
  "dialects": undefined,
  "en_example_sentence": undefined,
  "en_gloss_language": undefined,
  "es_example_sentence": undefined,
  "es_gloss_language": "árbol",
  "id": "34qw",
  "image_file_path": undefined,
  "image_filename": "",
  "interlinearization": undefined,
  "lexeme": "tree",
  "local_orthography_1": "𑃐𑃥𑃝𑃢 𑃒𑃦𑃗𑃠𑃤",
  "local_orthography_2": "চুড়া বংজি",
  "morphology": undefined,
  "notes": "",
  "noun_class": undefined,
  "parts_of_speech": undefined,
  "parts_of_speech_abbreviation": undefined,
  "phonetic": undefined,
  "plural_form": undefined,
  "semantic_domain_1": undefined,
  "semantic_domain_2": undefined,
  "sound_file_path": undefined,
  "sound_filename": "",
  "sources": undefined,
  "speaker_birthplace": undefined,
  "speaker_decade": undefined,
  "speaker_gender": undefined,
  "speaker_name": undefined,
  "vernacular_example_sentence": undefined,
}
`;
